### DISCUSSION TITLE
Net Meter Return to Grid and Solar Production Tracking

### DISCUSSION THREAD

**Discussion by Blacklabel7** [tronikos/opower#74] [Ideas]
I'm tracking consumption through Portland General Electric and my solar output is tracked with a Tesla integration.
Since I'm set up with net metering, I will see my consumption go negative when my solar output exceeds usage. 

If the negative values could be separated and entered into "Return to Grid" (as a positive value), the dashboard would accurately reflect how much energy is being consumed/generated over time. 
The first screenshot I should have total consumption 14.1 kWh and 2.5 kWh returned to grid. 
Second screenshot shows additional metrics in the energy dashboard when I enter my solar production in Return to Grid. 

This interest came up when I was trying to troubleshoot my solar panel output for time of day pricing. It's partly a cosmetic request, but it would really save me from doing the math myself. It could also be a big help with figuring out future battery upgrades to my house. I'm not dealing with NEM 3.0, but I imaging those customers would really appreciate this feature as well. 

<img width="700" alt="Screenshot 2024-03-19 at 5 07 50 PM" src="https://github.com/tronikos/opower/assets/14036697/b89a421a-8c08-44c7-b437-f2ee3759d05b">
<img width="714" alt="Screenshot 2024-03-21 at 7 59 43 AM" src="https://github.com/tronikos/opower/assets/14036697/607c763f-6a9d-44a6-8a56-006b326abff9">

**Comment by Blacklabel7** [comment]
I saw someone post a good example of how their energy dashboard looks. Having the positive and negative values separated provides more accurate results in the energy distribution section. 
![IMG_0817](https://github.com/tronikos/opower/assets/14036697/387e517f-2495-4b47-80e6-7726719ff96c)

**Comment by bloggrerishaq** [comment]
[Net metering](https://solarpanelrates.pk/net-metering/) allows solar panel owners to receive credit for excess electricity generated and sent back to the grid. When your solar output exceeds your consumption, your meter records negative usage. Separating these negative values as "Return to Grid" would improve your energy dashboard's accuracy, aid in troubleshooting solar output, and assist in planning battery upgrades

**Comment by tronikos** [comment]
I have solar too. It would be confusing to show 0 in the energy usage section of the dashboard when you return to the grid.

@Blacklabel7 @bloggrerishaq can you clarify how this would troubleshoot solar panel output when there is no way to know your usage? 

@Blacklabel7 @bloggrerishaq can you clarify how this would assist in planning battery upgrades?

**Comment by astavitsky** [comment]
I believe the issue is that the Home Assistant Energy Dashboard is not equipped to handle net metering from a single sensor. This is not unique to Opower integration. See [How can I configure the energy dashboard to understand my net electric meter?](https://community.home-assistant.io/t/how-can-i-configure-the-energy-dashboard-to-understand-my-net-electric-meter/383295)
m the grid was negative.

The solution presented in the linked thread is to use template sensors to create separate import and export sensors.

Which opower sensor should I use as a source for those template sensors?

**Reply by parkerbxyz** [reply]
The sensors available in the energy dashboard don't seem to be available to use with helpers. 😕

**Comment by parkerbxyz** [comment]
I would also like to have a separate sensor for use for the **return to grid** section.

In case it helps, I found a couple of similar requests in different communities:

- https://community.home-assistant.io/t/resolve-combine-multiple-energy-sources-in-energy-dashboard/625113/4

  > Just a bump in here after realizing another difficulty with the status quo-
Opower has one statistic that provides a net power consumption value over some time window.
There’s no explicit sensor for power returned to grid, which is also a bit annoying, since I can’t see it separately with the more granular data from my solar system without trashing the energy panel’s information.

- https://www.reddit.com/r/homeassistant/comments/18dvj2o/tesla_powerwall_solar_owners_with_pto_calculating/

  > I'm a new Tesla Powerwall owner with a solar inverter. Using the Tesla Powerwall Integration sensors, I can calculate how many kW's are sent back to the Grid daily.
  > 
  > However, I have no way of calculating credit $ for KW sent back to the power company.
  > 
  > In order to do that, I think I would need to convince the developer for OPower Integration to provide a new sensor that receives the current cost per kilowatt (based on the time of day and rate plan we subscribe to)... if available in the API.
  > 
  > Or... maybe just convincing the OPower developer to add a new sensor that receives current PTO credit so far.. if available in the API.
  > 
  > Can someone please confirm if the above 2 features for Opower would help me do what I want? Or, maybe I'm missing something?

**Comment by parkerbxyz** [comment]
I decided to take a crack at solving this. Here's what I came up with: https://github.com/home-assistant/core/pull/135258.

**Before:**

![before screenshot](https://github.com/user-attachments/assets/428dc5fc-6134-44a4-b995-f7933f07f69a)

**After:**

![after screenshot](https://github.com/user-attachments/assets/702f6457-599e-4ff2-9d51-fbd1f0ee7391)

**Comment by spencerthayer** [comment]
I'm looking for Home Assistant users who would like to get their PGE integration working again. I've gotten it working again via a HACS integration I've been coding this week. I've enabled discussion in the repo for conversations about the integration, I'm really looking for feedback and making this the best PGE solution for Home Assistant because Opower sucks.

https://github.com/spencerthayer/homeassistant-pge
